federation/pkg/federation-controller/service/servicecontroller.go

```diff
@@ -0,0 +1,830 @@
+		if err == nil || errors.IsNotFound(err) {
+			glog.V(4).Infof("service %s/%s deleted from cluster %s", service.Namespace, service.Name, clusterName)
+			return nil, !retryable
+		}
+		time.Sleep(clientRetryInterval)
+	}
+	glog.V(4).Infof("failed to delete service %s/%s from cluster %s, %+v", service.Namespace, service.Name, clusterName, err)
+	return err, !retryable
+}
+
+func (s *ServiceController) ensureClusterServices(clusterName string, service *v1.Service, client *clientset.Clientset) error {
+	var err error
+	var needUpdate bool
+	for i := 0; i < clientRetryCount; i++ {
+		svc, err := client.Core().Services(service.Namespace).Get(service.Name)
+		if err == nil && svc != nil {
+			// service exists
+			glog.V(5).Infof("found service %s/%s from cluster %s", service.Namespace, service.Name, clusterName)
+			// TODO, what properties can be updated by service controller?
+			// possibily:
+			// metadata.labels
+			// spec.type spec.selector spec.ports
+			if !reflect.DeepEqual(svc.Labels, service.Labels) {
+				svc.Labels = service.Labels
+				needUpdate = true
+			}
+			if !reflect.DeepEqual(svc.Spec.Type, service.Spec.Type) {
+				svc.Spec.Type = service.Spec.Type
+				needUpdate = true
+			}
+			if !reflect.DeepEqual(svc.Spec.Selector, service.Spec.Selector) {
+				svc.Spec.Selector = service.Spec.Selector
+				needUpdate = true
+			}
+			// TODO: make sure auto assigned node port is updated to federation apiserver, otherwise value of fed service port will be 0
+			if !portsEqualForLB(svc, service) {
+				needUpdate = true
+			}
+			if needUpdate {
+				_, err = client.Core().Services(svc.Namespace).UpdateStatus(svc)
+				if err == nil {
+					glog.V(5).Infof("service %s/%s successfully updated to cluster %s", svc.Namespace, svc.Name, clusterName)
+					return nil
+				} else {
+					glog.V(4).Infof("failed to update status %+v", err)
+				}
+			} else {
+				glog.V(5).Infof("service %s/%s is not updated to cluster %s as the spec are identical", svc.Namespace, svc.Name, clusterName)
+				return nil
+			}
+		}
+		if svc == nil || errors.IsNotFound(err) {
+			// Create service if it is not found
+			glog.Infof("Service '%s/%s' does not exist in cluster %s, trying to create new",
+				service.Namespace, service.Name, clusterName)
+			service.ResourceVersion = ""
+			_, err = client.Core().Services(service.Namespace).Create(service)
+			if err == nil {
+				glog.V(5).Infof("service %s/%s successfully created to cluster %s", service.Namespace, service.Name, clusterName)
+				return nil
+			}
+			glog.V(4).Infof("failed to create status %+v", err)
+		}
+		if errors.IsConflict(err) {
+			glog.V(4).Infof("Not persisting update to service '%s/%s' that has been changed since we received it: %v",
+				service.Namespace, service.Name, err)
+			// ignore and get LB info
+			return nil
+		}
+		glog.Warningf("Failed to update service '%s/%s' in cluster %s: %v",
```

Comment:
This will log even when the creation failed, and make the log message confusing.